src/storage/gc_worker.rs

```diff
@@ -408,6 +448,550 @@ impl<E: Engine> Runnable<GCTask> for GCRunner<E> {
+        }
+    }
+}
+
+/// The only error that will break `GCManager`'s process is that the `GCManager` is interrupted by
+/// others, maybe due to TiKV shutting down.
+#[derive(PartialEq, Debug)]
+enum GCManagerError {
+    Stopped,
+}
+
+type GCManagerResult<T> = ::std::result::Result<T, GCManagerError>;
+
+/// `GCManagerContext` is used to check if `GCManager` should be stopped.
+///
+/// When `GCManager` is running, it might take very long time to GC a round. It should be able to
+/// break at any time so that we can shut down TiKV in time.
+struct GCManagerContext {
+    /// Used to receive stop signal. The sender side is hold in `GCManagerHandler`.
+    /// If this field is `None`, the `GCManagerContext` will never stop.
+    stop_signal_receiver: Option<mpsc::Receiver<()>>,
+    /// Whether an stop signal is received.
+    is_stopped: bool,
+}
+
+impl GCManagerContext {
+    /// Creates a new `GCManagerContext` with no
+    pub fn new() -> Self {
+        Self {
+            stop_signal_receiver: None,
+            is_stopped: false,
+        }
+    }
+
+    /// Sets the receiver that used to receive the stop signal. `GCManagerContext` will be
+    /// considered to be stopped as soon as a message is received from the receiver.
+    pub fn set_stop_signal_receiver(&mut self, rx: mpsc::Receiver<()>) {
+        self.stop_signal_receiver = Some(rx);
+    }
+
+    /// Sleeps for a while. if a stop message is received, returns immediately with
+    /// `GCManagerError::Stopped`
+    fn sleep_or_stop(&mut self, timeout: Duration) -> GCManagerResult<()> {
+        if self.is_stopped {
+            return Err(GCManagerError::Stopped);
+        }
+        match self.stop_signal_receiver.as_ref() {
+            Some(rx) => match rx.recv_timeout(timeout) {
+                Ok(_) => {
+                    self.is_stopped = true;
+                    Err(GCManagerError::Stopped)
+                }
+                Err(mpsc::RecvTimeoutError::Timeout) => Ok(()),
+                Err(mpsc::RecvTimeoutError::Disconnected) => {
+                    panic!("stop_signal_receiver unexpectedly disconnected")
+                }
+            },
+            None => {
+                thread::sleep(timeout);
+                Ok(())
+            }
+        }
+    }
+
+    /// Checks if a stop message has been fired. Returns `GCManagerError::Stopped` if there's such
+    /// a message.
+    fn check_stopped(&mut self) -> GCManagerResult<()> {
+        if self.is_stopped {
+            return Err(GCManagerError::Stopped);
+        }
+        match self.stop_signal_receiver.as_ref() {
+            Some(rx) => match rx.try_recv() {
+                Ok(_) => {
+                    self.is_stopped = true;
+                    Err(GCManagerError::Stopped)
+                }
+                Err(mpsc::TryRecvError::Empty) => Ok(()),
+                Err(mpsc::TryRecvError::Disconnected) => {
+                    panic!("stop_signal_receiver unexpectedly disconnected")
```

Comment:
Panic will happen if the sender been closed unexpectedly, for example kill the tikv-server.